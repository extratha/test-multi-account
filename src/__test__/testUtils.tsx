import { RenderOptions, RenderResult, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SWRConfig } from 'swr';
import { theme } from '@/config/config-mui/theme';
import { ThemeProvider as MUIThemeProvider } from '@mui/material/styles';
import { NextIntlClientProvider } from 'next-intl';
import commonTh from '../../public/locales/th/common.json';
import commonEn from '../../public/locales/th/common.json';
type RenderWithProvider = (
  elm: React.ReactElement,
  renderOptions?: ProviderOptions,
) => RenderResult;

export type ProviderOptions = {
  formOptions?: Record<string, unknown>;
} & RenderOptions;

const messages = {
  ...commonTh,
  ...commonEn,
};

const renderWithProviders: RenderWithProvider = (
  component,
  renderOptions?,
  locale = 'th',
) => {
  const wrapper = ({ children }: { children: React.ReactNode }) => (
    <NextIntlClientProvider locale={locale} messages={messages}>
      <MUIThemeProvider theme={theme}>
        <SWRConfig value={{ provider: () => new Map() }}>{children}</SWRConfig>
      </MUIThemeProvider>
    </NextIntlClientProvider>
  );

  return render(component, {
    wrapper,
    ...renderOptions,
  });
};

export { renderWithProviders as render, userEvent };
export * from '@testing-library/react';
