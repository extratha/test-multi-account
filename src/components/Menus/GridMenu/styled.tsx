import { styled, Stack } from "@mui/material"

export const CustomMenuItemWrapperStyle = styled(Stack)(({ theme }) => ({
  border: `1px solid ${theme.palette.grey[400]}`,
  borderRadius: '20px',
  padding: '16px',
  justifyContent:'center',
  alignItems:'center',
  height:'100%',
  width:'100%',
}));