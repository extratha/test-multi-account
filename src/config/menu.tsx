import { ReactNode } from "react";

import { IconAiInterpret, IconHomePage, IconLogout, IconSetting } from "@/assets";

export const MENU_ICON: Record<string, ReactNode> = {
  HOME_ICON: <IconHomePage />,
  AI_MENU_ICON: <IconAiInterpret />,
  SETTING_ICON: <IconSetting />,
  LOGOUT_ICON: <IconLogout />,
};
