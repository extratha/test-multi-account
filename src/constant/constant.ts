export const GENDER = {
  MALE: "Male",
  FEMALE: "Female",
};

export const CONFIG_FIELD_TYPES = {
  STRING: "String",
  NUMBER: "Number",
  DROPDOWN: "Dropdown",
};
