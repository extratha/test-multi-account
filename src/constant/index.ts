export const ENV = {
  BASE_API_URL: process.env.NEXT_PUBLIC_BASE_API_URL || "MOCK_NEXT_PUBLIC_BASE_API_URL",
  FIREBASE_API_KEY: process.env.NEXT_PUBLIC_FIREBASE_API_KEY || "MOCK_NEXT_PUBLIC_FIREBASE_API_KEY",
  FIREBASE_AUTH_DOMAIN: process.env.NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN || "MOCK_NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN",
  FIREBASE_PROJECT_ID: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID || "MOCK_NEXT_PUBLIC_FIREBASE_PROJECT_ID",
  FIREBASE_STORAGE_BUCKET:
    process.env.NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET || "MOCK_NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET",
  FIREBASE_MESSAGING_SENDER_ID:
    process.env.NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID || "MOCK_NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID",
  FIREBASE_APP_ID: process.env.NEXT_PUBLIC_FIREBASE_APP_ID || "MOCK_NEXT_PUBLIC_FIREBASE_APP_ID",
  FIREBASE_MEASUREMENT_ID:
    process.env.NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID || "MOCK_NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID",
};

export const SESSION = {
  ACCESS_TOKEN: "access-token",
  REFRESH_TOKEN: "refresh-token",
  RESET_PASSWORD_TOKEN: "reset-password-token",
  X_ROBOT_TOKEN: "x-robot-token",
};

export const MENU_CONDITION = {
  LOGOUT: "logout",
};

export const NAVIGATION = {
  NOT_FOUND: "/not-found",
  LOGIN: "/login",
  HOME: "/home",
  FORGET_PASSWORD: "/forget-password",
  SET_NEW_PASSWORD: "/set-new-password",
  CONSENT_TERMS_CONDITIONS: "/consents/terms-conditions",
  CONSENT_PRIVACY_POLICIES: "/consents/privacy-policies",
  AI_INTERPRET_RESULT: "/ai-interpret/interpret",
  AI_INTERPRET_TRY_EXAMPLE_DATA: "/ai-interpret/try-example-data",
  AI_INTERPRET_TRY_INPUT_DATA: "/ai-interpret/try-input-data",
  SYMPTOM_CHECKER: "/symptom-checker",
  SETTINGS_TERMS_AND_CONDITIONS: "/settings/terms-and-conditions",
  SETTINGS_PRIVACY_POLICY: "/settings/privacy-policy",
  PUBLIC_TERMS_AND_CONDITIONS: "/public/consents/terms-conditions",
  PUBLIC_PRIVACY_POLICY: "/public/consents/privacy-policies",
};

export const GENDER = {
  MALE: "Male",
  FEMALE: "Female",
};

export const CONFIG_FIELD_TYPES = {
  STRING: "String",
  NUMBER: "Number",
  DROPDOWN: "Dropdown",
};

export const GROUP_NAME = {
  GENERAL_CHECK_UP: "generalCheckUp",
  HEMATOLOGY_BG: "hematologyBG",
  HEMATOLOGY_CBC: "hematologyCBC",
  LIPID_PROFILE: "lipidProfile",
  BLOOD_CHEMISTRY: "bloodChemistry",
  RENAL_FUNCTION_TEST: "renalFunctionTest",
  LIVER_FUNCTION_TEST: "liverFunctionTest",
  URINE_ANALYSIS: "urineAnalysis",
  STOOL_EXAMINATION: "stoolExamination",
  PAP_SMEAR: "papSmear",
  TUMOR_MARKER: "tumorMarker",
};

export const GENERAL_CHECK_UP = {
  AGE: "age",
  GENDER: "gender",
  BODY_WEIGHT: "body_weight",
  BODY_HEIGHT: "body_height",
  WAIST_LINE: "waist_line",
  BLOOD_PRESSURE_SYSTOLIC: "blood_pressure_systolic",
  BLOOD_PRESSURE_DIASTOLIC: "blood_pressure_diastolic",
  PULSE: "pulse",
  RESPIRATORY_RATE: "respiratory_rate",
  TEMPERATURE: "temperature",
};

export const HEMATOLOGY_BLOOD = {
  BLOOD_GROUP: "blood_group",
  BLOOD_GROUP_RH: "blood_group_rh",
};

export const HEMATOLOGY_BLOOD_GROUP = {
  HEMATOLOGY_BLOOD: "bloodGroup",
};

export const HEMATOLOGY_CBC = {
  HB: "hb_value",
  HCT: "hct_value",
  RBC_MORPHOLOGY: "rbc_morphology_value",
  MCV: "mcv_value",
  MCH: "mch_value",
  MCHC: "mchc_value",
  RDW: "rdw_value",
  WBC: "wbc_value",
  NEUTROPHIL: "neutrophil_value",
  LYMPHOCYTE: "lymphocyte_value",
  MONOCYTE: "monocyte_value",
  EOSINOPHIL: "eosinophil_value",
  BASOPHIL: "basophil_value",
  PLATELET_COUNT: "platelet_count_value",
};

export const HEMATOLOGY_CBC_GROUP = {
  HEMOGLOBIN: "hemoglobin",
  WHITE_BLOOD: "whiteBlood",
  RED_BLOOD: "redBlood",
};

export const INTERPRET_STATUS = {
  PENDING: "pending",
  SUCCESS: "success",
  FAILED: "failed",
};

export const ASR_SERVICE = {
  ORDER_ASR: "order-asr",
  SOAP_ASR: "soap-asr",
};

export const ASR_SERVICE_KEY: Record<string, string> = {
  [ASR_SERVICE.ORDER_ASR]: "orderAsr",
  [ASR_SERVICE.SOAP_ASR]: "soapAsr",
};

export const CONSENT_TYPE = {
  TERMS_CONDITIONS: "terms-conditions",
  PRIVACY_POLICIES: "privacy-policies",
};

export const SUBMIT_CONSENT_TYPE = {
  [CONSENT_TYPE.TERMS_CONDITIONS]: "terms",
  [CONSENT_TYPE.PRIVACY_POLICIES]: "privacy_policies",
};
