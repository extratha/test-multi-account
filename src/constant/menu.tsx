"use client";
import { IconHomePage, IconHomePageActive } from "@/assets";
import { webPaths } from "./webPaths";

export interface MenuItem {
  icon: any;
  activeIcon: any;
  path: null | string;
  title: null | string;
  children: (never | MenuItem)[] | null;
}
type TFunction = (key: string) => string;

export const iconSize = 24;
export const iconStyles = {
  width: iconSize,
  height: iconSize,
};

export const aiMenuList = (t: TFunction): MenuItem[] => {
  return [
    {
      icon: <IconHomePage style={iconStyles} />,
      activeIcon: <IconHomePageActive style={iconStyles} />,
      path: webPaths.home,
      title: t("Common.menu.home"),
      children: [],
    },
    {
      icon: null,
      activeIcon: null,
      path: webPaths.aiInterpret.tryExampleData,
      title: t("Common.menu.aiInterpret.title"),
      children: [
        {
          icon: null,
          activeIcon: null,
          path: webPaths.aiInterpret.tryExampleData,
          title: t("Common.menu.aiInterpret.tryExampleData"),
          children: null,
        },
        {
          icon: null,
          activeIcon: null,
          path: webPaths.aiInterpret.tryInputData,
          title: t("Common.menu.aiInterpret.tryInputData"),
          children: null,
        },
        {
          icon: null,
          activeIcon: null,
          path: webPaths.aiInterpret.tryHospitalData,
          title: t("Common.menu.aiInterpret.tryHospitalData"),
          children: null,
        },
      ],
    },
    // {
    //   icon: null,
    //   activeIcon: null,
    //   path: '',
    //   title: t('menu.healthVisualization.title'),
    //   children: [
    //     {
    //       icon: null,
    //       activeIcon: null,
    //       path: '',
    //       title: "",
    //       children: null
    //     }
    //   ],
    // },
    // {
    //   icon: null,
    //   activeIcon: null,
    //   path: '',
    //   title: t('menu.asr.title'),
    //   children: [
    //     {
    //       icon: null,
    //       activeIcon: null,
    //       path: '',
    //       title: "",
    //       children: null
    //     }
    //   ],
    // },
    {
      icon: null,
      activeIcon: null,
      path: webPaths.symptomChecker,
      title: t("Common.menu.symptomChecker.title"),
      children: [],
    },
  ];
};
