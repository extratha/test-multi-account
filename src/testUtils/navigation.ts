import * as NextNavigation from "next/navigation";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
  useParams: jest.fn(),
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
  useServerInsertedHTML: jest.fn(),
}));

export interface SpyUseSearchParams {
  get: jest.Mock;
}

export interface SpyUseRouter {
  replace: jest.Mock;
  push: jest.Mock;
  back: jest.Mock;
}

export const spyUseParams = (params: any = {}) => {
  jest.spyOn(NextNavigation, "useParams").mockReturnValue(params);
  return NextNavigation.useParams as jest.Mock;
};

export const spyUseRouter = (): SpyUseRouter => {
  const replace = jest.fn();
  const push = jest.fn();
  const back = jest.fn();
  jest.spyOn(NextNavigation, "useRouter").mockReturnValue({ replace, push, back } as any);
  return { replace, push, back };
};

export const spyUsePathname = (pathname: string) => {
  jest.spyOn(NextNavigation, "usePathname").mockReturnValue(pathname);
};

export const spyUseSearchParams = (): SpyUseSearchParams => {
  const get = jest.fn();
  jest.spyOn(NextNavigation, "useSearchParams").mockReturnValue({ get } as any);
  return {
    get,
  };
};

export const spyUseServerInsertedHTML = () => {
  return jest.spyOn(NextNavigation, "useServerInsertedHTML");
};
