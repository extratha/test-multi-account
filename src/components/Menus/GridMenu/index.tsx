'use client'

import { Grid, MenuClassKey, Stack, Typography, useTheme } from "@mui/material";
import { MenuItem } from "@/constant/menu";
import Image from "next/image";
import { IconMenuItem } from "@/assets";
import { CustomMenuItemWrapperStyle } from "./styled";
import { useTranslations } from "next-intl";


const GridMenu: React.FC<{ menus: MenuItem[] }> = ({ menus }) => {
  const theme = useTheme()
  const t = useTranslations('Common')
  const MenuRenderer = () => {
    const iconHeight = 48
    return (
      <>
        {menus.map((menu, index) => {
          return (
            <Grid 
            item 
            xs={4} 
            sx={{ 
              display: 'flex', 
              width: '100%', 
              height: '300px' 
            }}
            key={index}
            >
              <CustomMenuItemWrapperStyle spacing={2}>
                <Image alt='' src={IconMenuItem} style={{ width: iconHeight, height: iconHeight, margin: "0 0 8px 0" }} />
                <Typography variant='titleLargeSemiBold' textAlign="center">{menu.title}</Typography>
                <Typography variant='bodyLarge' color={theme.palette.grey[600]} >{t('text.interpret')}</Typography>
              </CustomMenuItemWrapperStyle>
            </Grid>
          )
        })}
      </>
    )
  }
  return (
    <>
      <Grid container spacing={3} alignItems={'stretch'}>
        {MenuRenderer()}
      </Grid>
    </>
  )
}

export default GridMenu;