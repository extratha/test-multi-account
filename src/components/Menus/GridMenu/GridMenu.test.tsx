import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { aiMenuList } from "@/constant/menu";
import { spyUseRouter } from "@/testUtils/navigation";
import { render } from "@/testUtils/testUtils";
import GridMenu from ".";

describe("GridMenu Component", () => {
  const mockTFunction = (key: string) => key;

  beforeEach(() => {
    spyUseRouter();
  });

  const renderGridMenu = () => {
    return render(<GridMenu menus={aiMenuList(mockTFunction)} />);
  };

  it("should render correctly", async () => {
    const { asFragment } = renderGridMenu();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should call router push when click menu item", async () => {
    renderGridMenu();

    await userEvent.click(screen.getByTestId("menu-item-0"));
    // TODO: expect router push with ?
  });
});
