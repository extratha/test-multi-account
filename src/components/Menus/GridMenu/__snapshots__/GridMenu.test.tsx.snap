// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridMenu Component should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-top: -24px;
  width: calc(100% + 24px);
  margin-left: -24px;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.emotion-0>.MuiGrid-item {
  padding-top: 24px;
}

.emotion-0>.MuiGrid-item {
  padding-left: 24px;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-basis: 33.333333%;
  -ms-flex-preferred-size: 33.333333%;
  flex-basis: 33.333333%;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  max-width: 33.333333%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 260px;
}

@media (min-width:600px) {
  .emotion-1 {
    -webkit-flex-basis: 33.333333%;
    -ms-flex-preferred-size: 33.333333%;
    flex-basis: 33.333333%;
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    max-width: 33.333333%;
  }
}

@media (min-width:900px) {
  .emotion-1 {
    -webkit-flex-basis: 33.333333%;
    -ms-flex-preferred-size: 33.333333%;
    flex-basis: 33.333333%;
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    max-width: 33.333333%;
  }
}

@media (min-width:1200px) {
  .emotion-1 {
    -webkit-flex-basis: 33.333333%;
    -ms-flex-preferred-size: 33.333333%;
    flex-basis: 33.333333%;
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    max-width: 33.333333%;
  }
}

@media (min-width:1536px) {
  .emotion-1 {
    -webkit-flex-basis: 33.333333%;
    -ms-flex-preferred-size: 33.333333%;
    flex-basis: 33.333333%;
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    max-width: 33.333333%;
  }
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: 1px solid #bdbdbd;
  border-radius: 20px;
  padding: 16px;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  width: 100%;
  cursor: pointer;
}

.emotion-2>:not(style):not(style) {
  margin: 0;
}

.emotion-2>:not(style)~:not(style) {
  margin-top: 4px;
}

.emotion-3 {
  margin: 0;
  font-size: 16px;
  font-weight: 700;
  line-height: 1.5;
  text-align: center;
}

.emotion-4 {
  margin: 0;
  font-size: 14px;
  line-height: 1.5;
  color: #607d8b;
}

<div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 emotion-0"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 emotion-1"
      data-testid="menu-item-0"
    >
      <div
        class="MuiStack-root emotion-2"
      >
        <svg
          style="width: 48px; height: 48px; margin: 0px 0px 16px 0px;"
        />
        <p
          class="MuiTypography-root MuiTypography-bodyBold emotion-3"
        >
          Common.menu.home
        </p>
        <p
          class="MuiTypography-root MuiTypography-bodySmall emotion-4"
        >
          แปลและสรุปผลสุขภาพ
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 emotion-1"
      data-testid="menu-item-1"
    >
      <div
        class="MuiStack-root emotion-2"
      >
        <svg
          style="width: 48px; height: 48px; margin: 0px 0px 16px 0px;"
        />
        <p
          class="MuiTypography-root MuiTypography-bodyBold emotion-3"
        >
          Common.menu.aiInterpret.title
        </p>
        <p
          class="MuiTypography-root MuiTypography-bodySmall emotion-4"
        >
          แปลและสรุปผลสุขภาพ
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 emotion-1"
      data-testid="menu-item-2"
    >
      <div
        class="MuiStack-root emotion-2"
      >
        <svg
          style="width: 48px; height: 48px; margin: 0px 0px 16px 0px;"
        />
        <p
          class="MuiTypography-root MuiTypography-bodyBold emotion-3"
        >
          Common.menu.symptomChecker.title
        </p>
        <p
          class="MuiTypography-root MuiTypography-bodySmall emotion-4"
        >
          แปลและสรุปผลสุขภาพ
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
